homeassistant/components/dlna_dmr/config_flow.py

```diff
@@ -93,8 +93,9 @@ async def async_step_user(self, user_input: FlowInput = None) -> FlowResult:
             return await self.async_step_manual()
 
         self._discoveries = {
-            discovery.get(ssdp.ATTR_UPNP_FRIENDLY_NAME)
-            or urlparse(discovery[ssdp.ATTR_SSDP_LOCATION]).hostname: discovery
+            discovery.upnp.get(ssdp.ATTR_UPNP_FRIENDLY_NAME)
+            or urlparse(discovery.ssdp_location or "").hostname
+            or "": discovery
             for discovery in discoveries
         }
```

Comment:
I don't think a discovery is valid for `dlna_dmr` if it doesn't have a location.

Reply:
This is a workaround due to `mypy`, see this comment from `hue` PR: https://github.com/home-assistant/core/pull/60598#discussion_r759258314

Do we need to add a global helper to extract the hostname from an ssdp_location, raising a `ValueError` if either `ssdp_location` or resulting `hostname` is None?

Reply:
I have added a local helper (not covered by tests yet).
The other option is to add a `#type-ignore` but I don't think that would be best.

Reply:
As noted by @bdraco, dlna_dmr is not interested in discoveries without a location, so a smaller solution would be:
```py
        self._discoveries = {
            discovery.upnp.get(ssdp.ATTR_UPNP_FRIENDLY_NAME)
            or cast(str, urlparse(discovery.ssdp_location).hostname): discovery
            for discovery in discoveries
            if discovery.ssdp_location
        }
```

The cast is necessary because mypy can't figure out that `urlparse` will return a `ParseResult` and not `ParseResultBytes` when it's given an `str`.

Reply:
I offered the solution with an external `_parse_hostname` helper because I have bumped into this issue on multiple occasions.
I also had complaints about the `cast` solution from other reviewers.
For the moment I have used `type ignore` in #60787 but I am not convinced it's the best solution.

Reply:
cc @MartinHjelmare

Reply:
I have used the solution recommended by @chishm.
I have created a separate PR for a `parse_hostname` solution: #60839